rllib/examples/learners/classes/mog_loss.py

```diff
@@ -0,0 +1,183 @@
+        )
+        # little trick of using log_softmax on the current alphas to prevent nans
+        loga = torch.clamp(nn.functional.log_softmax(alpha_current, dim=-1), 1e-6, None)
+
+        summing_log = -torch.logsumexp(logp + loga, dim=-1)
+        return summing_log
+
+    @override(PPOTorchLearner)
+    def compute_loss_for_module(
+        self,
+        *,
+        module_id: ModuleID,
+        config: CustomPPOConfig,
+        batch: Dict[str, Any],
+        fwd_out: Dict[str, TensorType],
+    ) -> TensorType:
+
+        if Columns.LOSS_MASK in batch:
+            mask = batch[Columns.LOSS_MASK]
+            num_valid = torch.sum(mask)
+
+            def possibly_masked_mean(data_):
+                return torch.sum(data_[mask]) / num_valid
+
+        else:
+            possibly_masked_mean = torch.mean
+
+        module = self.module[module_id].unwrapped()
+
+        # batch in the learner is: SampleBatch(400: ['loss_mask', 'terminateds', 'obs', 'actions', 'rewards', 'truncateds',
+        # 'action_dist_inputs', 'value_function_out', 'mog_components', 'action_logp', 'weights_seq_no', 'advantages', 'value_targets', 'infos'])
+
+        action_dist_class_train = module.get_train_action_dist_cls()
+        action_dist_class_exploration = module.get_exploration_action_dist_cls()
+
+        curr_action_dist = action_dist_class_train.from_logits(
+            fwd_out[Columns.ACTION_DIST_INPUTS]
+        )
+        prev_action_dist = action_dist_class_exploration.from_logits(
+            batch[Columns.ACTION_DIST_INPUTS]
+        )
+
+        logp_ratio = torch.exp(
+            curr_action_dist.logp(batch[Columns.ACTIONS]) - batch[Columns.ACTION_LOGP]
+        )
+
+        # Only calculate kl loss if necessary (kl-coeff > 0.0).
+        if config.use_kl_loss:
+            action_kl = prev_action_dist.kl(curr_action_dist)
+            mean_kl_loss = possibly_masked_mean(action_kl)
+        else:
+            mean_kl_loss = torch.tensor(0.0, device=logp_ratio.device)
+
+        curr_entropy = curr_action_dist.entropy()
+        mean_entropy = possibly_masked_mean(curr_entropy)
+
+        surrogate_loss = torch.min(
+            batch[Postprocessing.ADVANTAGES] * logp_ratio,
+            batch[Postprocessing.ADVANTAGES]
+            * torch.clamp(logp_ratio, 1 - config.clip_param, 1 + config.clip_param),
+        )
+
+        # Compute the MOG value loss using the negative log-likelihood
+        # source of loss function @ricefield
+        if config.use_critic:
+            rewards = batch["rewards"]
+            dones = batch["dones"]
+            gamma = config["gamma"]
+
+            mog_components = fwd_out["mog_components"]
+            mu_current = mog_components["means"]
+            sigmas_current = mog_components["sigmas"]
+            alpha_current = mog_components["alphas"]
+
+            # pass through the network
+            if "new_obs" not in batch:
```

Comment:
I would like to propose to change this to either no check at all - you are forcing this actually by the custom `AlgorithmConfig` class - or to raise an error, if the `NEXT_OBS` are not in the batch.

Reply:
Small nit: We have constants for all components of the batch: 

```python
from ray.rllib.core.columns import Columns

batch[Columns.NEXT_OBS]
```

You could provide here new constants `MOG_COMPONENTS`, ...